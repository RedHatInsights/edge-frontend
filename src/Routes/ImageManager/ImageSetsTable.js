import React from 'react';
import GeneralTable from '../../components/general-table/GeneralTable';
import PropTypes from 'prop-types';
import { routes as paths } from '../../constants/routeMapper';
import { createLink } from '../../utils';
import { Tooltip } from '@patternfly/react-core';
import { DateFormat } from '@redhat-cloud-services/frontend-components/DateFormat';
import { cellWidth } from '@patternfly/react-table';
import CustomEmptyState from '../../components/Empty';
import { useLocation, useHistory, useNavigate } from 'react-router-dom';
import { emptyStateNoFilters } from '../../utils';
import Status from '../../components/Status';
import { getBaseURLFromPrefixAndName } from '../ImageManagerDetail/utils';
import { distributionMapper } from '../../constants';

const TooltipSelectorRef = ({ index }) => (
  <div>
    <Tooltip
      content={<div>More options</div>}
      reference={() =>
        document.getElementById(`pf-dropdown-toggle-id-${index}`)
      }
    />
  </div>
);

TooltipSelectorRef.propTypes = {
  index: PropTypes.number,
};

const defaultFilters = [
  {
    label: 'Name',
    type: 'text',
  },
  {
    label: 'Status',
    type: 'checkbox',
    options: [
      { option: 'Building', value: 'BUILDING' },
      { option: 'Created', value: 'CREATED' },
      { option: 'Error', value: 'ERROR' },
      { option: 'Ready', value: 'SUCCESS' },
    ],
  },
];

const columnNames = [
  {
    title: 'Name',
    type: 'name',
    sort: true,
    columnTransforms: [cellWidth(15)],
  },
  {
    title: 'Version',
    type: 'version',
    sort: false,
    columnTransforms: [cellWidth(15)],
  },
  {
    title: 'Release',
    type: 'distribution',
    options: distributionMapper,
    sort: false,
    columnTransforms: [cellWidth(15)],
  },
  {
    title: 'Target',
    type: 'outputTypes',
    sort: false,
    columnTransforms: [cellWidth(15)],
  },
  {
    title: 'Status',
    type: 'status',
    sort: false,
    columnTransforms: [cellWidth(15)],
  },
  {
    title: 'Created/Updated',
    type: 'updated_at',
    sort: true,
    columnTransforms: [cellWidth(35)],
  },
];
const createRows = (data, baseURL, history, navigate) => {
  return data.map((image_set, index) => ({
    rowInfo: {
      id: image_set?.ID,
      imageStatus: image_set?.Status,
      distribution: image_set?.Distribution,
      outputType: image_set?.OutputTypes,
      isoURL: image_set?.ImageBuildIsoURL || null,
      latestImageID: image_set?.ImageID,
    },
    cells: [
      {
        title: createLink({
          pathname: `${baseURL}/${image_set?.ID}`,
          linkText: image_set?.Name,
          history,
          navigate,
        }),
      },

      image_set?.Version,
      {
        title: distributionMapper[image_set?.Distribution],
      },
      image_set?.OutputTypes.length == 2
        ? 'Bare Metal Installer'
        : 'Update Only',
      {
        title: (
          <>
            {/* workaround for tooltip on kebab*/}
            <TooltipSelectorRef index={index} />
            <Status type={image_set?.Status.toLowerCase()} />
          </>
        ),
      },
      {
        title: image_set?.UpdatedAt ? (
          <DateFormat date={image_set?.UpdatedAt} />
        ) : (
          'Unknown'
        ),
      },
    ],
  }));
};

const ImageTable = ({
  pathPrefix,
  urlName,
  historyProp,
  locationProp,
  navigateProp,
  data,
  count,
  isLoading,
  hasError,
  fetchImageSets,
  openCreateWizard,
  openUpdateWizard,
  hasModalSubmitted,
  setHasModalSubmitted,
}) => {
  const { search } = locationProp
    ? locationProp()
    : useLocation
    ? useLocation()
    : null;
  const history = historyProp
    ? historyProp()
    : useHistory
    ? useHistory()
    : null;
  const navigate = navigateProp
    ? navigateProp()
    : useNavigate
    ? useNavigate()
    : null;
  const actionResolver = (rowData) => {
    const actionsArray = [];
    if (rowData.rowInfo?.isoURL) {
      actionsArray.push({
        title: 'Download',
        onClick: (_event, _rowId, rowData) => {
          window.open(rowData.rowInfo?.isoURL);
        },
      });
    }

    if (
      rowData.rowInfo?.imageStatus === 'SUCCESS' ||
      rowData.rowInfo?.imageStatus === 'ERROR'
    ) {
      actionsArray.push({
        title: 'Update Image',
        onClick: (_event, _rowId, rowData) => {
          openUpdateWizard(rowData.rowInfo?.latestImageID);
        },
      });
    }

    if (rowData.rowInfo?.imageStatus === 'BUILDING' && rowData.rowInfo?.id) {
      actionsArray.push({
        title: '',
      });
    }

    return actionsArray;
  };

  const areActionsDisabled = (rowData) =>
    rowData.rowInfo?.imageStatus === 'BUILDING';

  const baseURL = getBaseURLFromPrefixAndName(
    paths.manageImages,
    pathPrefix,
    urlName
  );

  return (
    <>
      {emptyStateNoFilters(isLoading, count, search) ? (
        <CustomEmptyState
          data-testid="general-table-empty-state-no-data"
          icon={'plus'}
          title={'Create an OSTree image'}
          body={
            'Use the image builder tool to compose ' +
            'customized RHEL (rpm-ostree) images optimized for Edge.' +
            ' Then, you can remotely install, and securely manage and' +
            ' scale deployments of the images on Edge servers.'
          }
          primaryAction={{
            click: openCreateWizard,
            text: 'Create new image',
          }}
          secondaryActions={[
            {
              type: 'link',
              title: 'RHEL for Edge image documentation',
              link: 'https://access.redhat.com/documentation/en-us/edge_management/2022/html/create_rhel_for_edge_images_and_configure_automated_management/index',
            },
          ]}
        />
      ) : (
        <GeneralTable
          apiFilterSort={true}
          historyProp={historyProp}
          locationProp={locationProp}
          navigateProp={navigateProp}
          isUseApi={true}
          filters={defaultFilters}
          loadTableData={fetchImageSets}
          tableData={{ count, data, isLoading, hasError }}
          columnNames={columnNames}
          rows={data ? createRows(data, baseURL, history, navigate) : []}
          actionResolver={actionResolver}
          areActionsDisabled={areActionsDisabled}
          defaultSort={{ index: 5, direction: 'desc' }}
          toolbarButtons={[
            {
              title: 'Create new image',
              click: () => openCreateWizard(),
            },
          ]}
          hasModalSubmitted={hasModalSubmitted}
          setHasModalSubmitted={setHasModalSubmitted}
        />
      )}
    </>
  );
};

ImageTable.propTypes = {
  pathPrefix: PropTypes.string,
  urlName: PropTypes.string,
  historyProp: PropTypes.func,
  locationProp: PropTypes.func,
  navigateProp: PropTypes.func,
  data: PropTypes.array,
  count: PropTypes.number,
  isLoading: PropTypes.bool,
  hasError: PropTypes.bool,
  fetchImageSets: PropTypes.func,
  clearFilters: PropTypes.func,
  openCreateWizard: PropTypes.func,
  openUpdateWizard: PropTypes.func,
  filters: PropTypes.array,
  pagination: PropTypes.shape({
    page: PropTypes.number,
    perPage: PropTypes.number,
  }),
  hasModalSubmitted: PropTypes.bool,
  setHasModalSubmitted: PropTypes.func,
};

export default ImageTable;
