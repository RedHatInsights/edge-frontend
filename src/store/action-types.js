import keyBy from 'lodash/keyBy';
import flatMap from 'lodash/flatMap';

export const LOAD_DEVICE_SUMMARY = 'LOAD_DEVICE_SUMMARY';
export const LOAD_ACTIVE_IMAGES = 'LOAD_ACTIVE_IMAGES';
export const LOAD_EDGE_IMAGES = 'LOAD_EDGE_IMAGES';
export const LOAD_EDGE_IMAGE_SETS = 'LOAD_EDGE_IMAGE_SETS';
export const LOAD_EDGE_IMAGE_PACKAGES = 'LOAD_EDGE_IMAGE_PACKAGES';
export const LOAD_GROUPS = 'LOAD_GROUPS';
export const LOAD_GROUP_DETAIL = 'LOAD_GROUP_DETAIL';
export const LOAD_TRESHOLD = 'LOAD_TRESHOLD';
export const LOAD_DEVICES_INFO = 'LOAD_DEVICES_INFO';
export const LOAD_CANARIES_INFO = 'LOAD_CANARIES_INFO';
export const LOAD_GROUP_DEVICES_INFO = 'LOAD_GROUP_DEVICES_INFO';
export const LOAD_IMAGE_STATUS = 'LOAD_IMAGE_STATUS';
export const LOAD_IMAGE_DETAIL = 'LOAD_IMAGE_DETAIL';
export const LOAD_IMAGE_SET_DETAIL = 'LOAD_IMAGE_SET_DETAIL';
export const CREATE_NEW_IMAGE = 'CREATE_NEW_IMAGE';
export const CREATE_NEW_IMAGE_RESET = `${CREATE_NEW_IMAGE}_RESET`;
export const POLLING_IMAGES = 'POLLING_IMAGES';
export const LOAD_DEVICE_TABLE = 'LOAD_DEVICE_TABLE';
export const UPDATE_DISPLAY_NAME = 'UPDATE_DISPLAY_NAME';
export const REMOVE_ENTITY = 'REMOVE_ENTITY';

const asyncActions = flatMap(
  [
    LOAD_EDGE_IMAGES,
    LOAD_EDGE_IMAGE_SETS,
    LOAD_EDGE_IMAGE_PACKAGES,
    LOAD_ACTIVE_IMAGES,
    LOAD_GROUPS,
    LOAD_GROUP_DETAIL,
    LOAD_TRESHOLD,
    LOAD_DEVICES_INFO,
    LOAD_CANARIES_INFO,
    LOAD_GROUP_DEVICES_INFO,
    LOAD_IMAGE_STATUS,
    LOAD_IMAGE_DETAIL,
    LOAD_IMAGE_SET_DETAIL,
    CREATE_NEW_IMAGE,
    LOAD_DEVICE_TABLE,
    UPDATE_DISPLAY_NAME,
    REMOVE_ENTITY,
  ],
  (a) => [a, `${a}_PENDING`, `${a}_FULFILLED`, `${a}_REJECTED`]
);
export const ACTION_TYPES = keyBy(asyncActions, (k) => k);

export const SELECT_ENTITY = 'SELECT_ENTITY';
export const PRE_SELECT_ENTITY = 'PRE_SELECT_ENTITY';
export const CLEAN_ENTITIES = 'CLEAN_ENTITIES';
