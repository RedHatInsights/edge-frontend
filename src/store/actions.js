import {
  ACTION_TYPES,
  LOAD_TRESHOLD,
  LOAD_DEVICES_INFO,
  LOAD_CANARIES_INFO,
  LOAD_GROUP_DETAIL,
  LOAD_GROUP_DEVICES_INFO,
  SELECT_ENTITY,
  PRE_SELECT_ENTITY,
  CLEAN_ENTITIES,
  LOAD_ACTIVE_IMAGES,
  LOAD_EDGE_IMAGES,
  LOAD_EDGE_IMAGE_SETS,
  LOAD_EDGE_IMAGE_PACKAGES,
  LOAD_DEVICE_SUMMARY,
  LOAD_IMAGE_STATUS,
  LOAD_IMAGE_DETAIL,
  LOAD_IMAGE_SET_DETAIL,
  CREATE_NEW_IMAGE,
  POLLING_IMAGES,
  LOAD_DEVICE_TABLE,
} from './action-types';

import {
  fetchImageStatus,
  getImageById,
  fetchEdgeImages,
  fetchEdgeImageSets,
  getImagePackageMetadata,
  createImage,
  getImageSet,
  fetchActiveImages,
} from '../api/images';
import { getInventory } from '../api/devices';
import { hosts } from '../api';
import { deleteSystemsById } from '../utils';

export const loadThreshold = () => ({
  type: LOAD_TRESHOLD,
  payload: () => {},
});

export const loadDevicesInfo = () => ({
  type: LOAD_DEVICES_INFO,
  payload: () => {},
});

export const loadCanariesInfo = () => ({
  type: LOAD_CANARIES_INFO,
  payload: () => {},
});

export const loadGroupsDetail = () => ({
  type: LOAD_GROUP_DETAIL,
  payload: () => {},
});

export const loadGroupDevicesInfo = () => ({
  type: LOAD_GROUP_DEVICES_INFO,
  payload: () => {},
});

export const selectEntity = (id, selected) => ({
  type: SELECT_ENTITY,
  payload: {
    id,
    selected,
  },
});

export const preSelectEntity = (id, selected) => ({
  type: PRE_SELECT_ENTITY,
  payload: {
    id,
    selected,
  },
});

export const cleanEntities = () => ({
  type: CLEAN_ENTITIES,
});

export const loadImages = (dispatch, pagination) => {
  dispatch({
    type: LOAD_ACTIVE_IMAGES,
    payload: fetchActiveImages(pagination),
    meta: {
      notifications: {
        rejected: {
          variant: 'danger',
          title: 'Cannot show images data',
          description: 'Failed receiving images from image-builder',
        },
      },
    },
    // the '.catch' part is necessary because redux-promise-middleware throws the error on REJECTED
    // and to avoid the app exploding I need to catch it here.
    // THANK you redux-promise-middleware for not allowing to customize this behavior. ðŸ˜ 
  }).catch(() => null);
};

export const loadDeviceSummary = (dispatch) => {
  dispatch({
    type: LOAD_DEVICE_SUMMARY,
    payload: () => {},
    meta: {
      notifications: {
        rejected: {
          variant: 'danger',
          title: 'Cannot show system summary data',
          description: 'Failed receiving system summary data from inventory',
        },
      },
    },
    // the '.catch' part is necessary because redux-promise-middleware throws the error on REJECTED
    // and to avoid the app exploding I need to catch it here.
    // THANK you redux-promise-middleware for not allowing to customize this behavior. ðŸ˜ 
  }).catch(() => null);
};

export const loadImageStatus = (dispatch, imageId) => {
  dispatch({
    type: LOAD_IMAGE_STATUS,
    payload: fetchImageStatus({ id: imageId }),
  }).catch(() => null);
};

export const loadImageDetail = (dispatch, imageId) => {
  dispatch({
    type: LOAD_IMAGE_DETAIL,
    payload: getImageById({ id: imageId }),
  }).catch(() => null);
};

export const createNewImage = (dispatch, payload, callback) => {
  dispatch({
    type: CREATE_NEW_IMAGE,
    payload: createImage(payload),
  })
    .then(callback)
    .catch(() => null);
};

export const loadEdgeImages = (dispatch, query) => {
  dispatch({
    type: LOAD_EDGE_IMAGES,
    payload: fetchEdgeImages(query),
  }).catch(() => null);
};

export const loadEdgeImageSets = (dispatch, query) => {
  dispatch({
    type: LOAD_EDGE_IMAGE_SETS,
    payload: fetchEdgeImageSets(query),
  }).catch(() => null);
};

export const setPolling = (toStart, interval) => {
  const subAction = toStart ? 'START' : 'END';
  const payload = toStart ? { interval } : {};
  return {
    type: `${POLLING_IMAGES}_${subAction}`,
    ...payload,
  };
};

export const addImageToPoll = ({ id, name }) => {
  return {
    type: `${POLLING_IMAGES}_ADD`,
    payload: {
      name,
      id,
    },
  };
};

export const removeImagesToPoll = (ids) => {
  return {
    type: `${POLLING_IMAGES}_REMOVE`,
    ids,
  };
};

export const loadImageSetDetail = (dispatch, urlParam, query) => {
  dispatch({
    type: LOAD_IMAGE_SET_DETAIL,
    payload: getImageSet({ id: urlParam, q: query }),
  }).catch(() => null);
};

export const loadImagePackageMetadata = (dispatch, imageId) => {
  dispatch({
    type: LOAD_EDGE_IMAGE_PACKAGES,
    payload: getImagePackageMetadata(imageId),
  }).catch(() => null);
};

export const loadDeviceTable = (dispatch) => {
  dispatch({
    type: LOAD_DEVICE_TABLE,
    payload: getInventory(),
  }).catch(() => null);
};

export const editDisplayName = (id, value, origValue) => ({
  type: ACTION_TYPES.UPDATE_DISPLAY_NAME,
  payload: hosts.apiHostPatchHostById([id], { display_name: value }),
  meta: {
    id,
    value,
    origValue,
    notifications: {
      fulfilled: {
        variant: 'success',
        title: `Display name for entity with ID ${id} has been changed to ${value}`,
        dismissable: true,
      },
    },
  },
});

export const deleteEntity = (systems, displayName) => ({
  type: ACTION_TYPES.REMOVE_ENTITY,
  payload: deleteSystemsById(systems),
  meta: {
    notifications: {
      fulfilled: {
        variant: 'success',
        title: 'Delete operation finished',
        description: `${displayName} has been successfully removed.`,
        dismissable: true,
      },
    },
    systems,
  },
});
