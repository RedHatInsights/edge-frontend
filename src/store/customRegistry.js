import ReducerRegistry from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';

export const createRegistry = (
  initialState = {},
  middleware = [],
  composeEnhancersDefault
) => {
  return new ReducerRegistry(
    initialState,
    [...middleware],
    composeEnhancersDefault
  );
};
