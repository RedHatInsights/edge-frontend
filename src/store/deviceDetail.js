import GeneralInformationTab from '../components/DeviceDetail';
import { applyReducerHash } from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';

const entityLoaded = (state) => {
  return {
    ...state,
    loaded: true,
    activeApps: [
      {
        title: 'General information',
        name: 'general_information',
        component: GeneralInformationTab,
      },
    ],
  };
};

export const deviceDetail = applyReducerHash({
  LOAD_ENTITY_FULFILLED: entityLoaded,
});
