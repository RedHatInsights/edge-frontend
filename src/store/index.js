import { createContext } from 'react';
import { ReducerRegistry } from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';
import promiseMiddleware from 'redux-promise-middleware';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications/notificationsMiddleware';
import { polling } from './polling';

export const RegistryContext = createContext({
  getRegistry: () => {},
});

export function init(...middleware) {
  return new ReducerRegistry({}, [
    promiseMiddleware,
    notificationsMiddleware({
      errorDescriptionKey: ['detail', 'stack'],
    }),
    polling(),
    ...middleware,
  ]);
}
