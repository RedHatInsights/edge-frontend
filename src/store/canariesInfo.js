import { applyReducerHash } from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';
import { LOAD_CANARIES_INFO } from './action-types';

const initialState = {};

const loadCanariesInfoPending = (state) => ({
  ...state,
  canariesInfo: [],
  isLoading: true,
});
const loadCanariesInfoFulfilled = (state, { payload }) => ({
  ...state,
  canariesInfo: payload?.results || [],
  isLoading: false,
});

export default applyReducerHash(
  {
    [`${LOAD_CANARIES_INFO}_PENDING`]: loadCanariesInfoPending,
    [`${LOAD_CANARIES_INFO}_FULFILLED`]: loadCanariesInfoFulfilled,
  },
  initialState
);
