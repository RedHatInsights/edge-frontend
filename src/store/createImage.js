import { applyReducerHash } from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';
import { CREATE_NEW_IMAGE } from './action-types';

const initialState = {};

const createNewImagePending = () => ({
  isLoading: true,
  hasError: false,
  error: null,
});

const createNewImageFulfilled = () => {
  return {
    isLoading: false,
    hasError: false,
    error: null,
  };
};

const createNewImageRejected = (_state, { payload }) => ({
  isLoading: false,
  hasError: true,
  error: payload,
});

const createNewImageReset = () => ({
  isLoading: false,
  hasError: false,
  error: null,
});

export default applyReducerHash(
  {
    [`${CREATE_NEW_IMAGE}_PENDING`]: createNewImagePending,
    [`${CREATE_NEW_IMAGE}_FULFILLED`]: createNewImageFulfilled,
    [`${CREATE_NEW_IMAGE}_REJECTED`]: createNewImageRejected,
    [`${CREATE_NEW_IMAGE}_RESET`]: createNewImageReset,
  },
  initialState
);
