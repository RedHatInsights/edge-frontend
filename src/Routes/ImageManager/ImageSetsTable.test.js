import React from 'react';
import ImageSetsTable from './ImageSetsTable';
import { Provider } from 'react-redux';
import { init, RegistryContext } from '../../store';
import { render } from '@testing-library/react';
import logger from 'redux-logger';
import { MemoryRouter } from 'react-router-dom';

describe('ImageSets table', () => {
  it('to render correctly', () => {
    const openCreateWizard = jest.fn();
    const openUpdateWizard = jest.fn();
    const registry = init(logger);

    const { container } = render(
      <RegistryContext.Provider
        value={{
          getRegistry: () => registry,
        }}
      >
        <Provider store={registry.getStore()}>
          <ImageSetsTable
            openCreateWizard={openCreateWizard}
            openUpdateWizard={openUpdateWizard}
          />
        </Provider>
      </RegistryContext.Provider>,
      { wrapper: MemoryRouter }
    );

    expect(container.querySelector('div')).toMatchSnapshot();
  });
});
