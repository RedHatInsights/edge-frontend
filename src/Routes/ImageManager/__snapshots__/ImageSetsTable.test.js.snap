// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageSets table to render correctly 1`] = `
<div
  class="pf-v5-c-empty-state pf-m-lg"
>
  <div
    class="pf-v5-c-empty-state__content"
  >
    <div
      class="pf-v5-c-empty-state__icon"
    >
      <svg
        aria-hidden="true"
        class="pf-v5-svg"
        fill="currentColor"
        height="1em"
        role="img"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm144 276c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92h-92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
        />
      </svg>
    </div>
    <div
      class="pf-v5-c-empty-state__header"
    >
      <div
        class="pf-v5-c-empty-state__title"
      >
        <h4
          class="pf-v5-c-empty-state__title-text"
        >
          Create an OSTree image
        </h4>
      </div>
    </div>
    <div
      class="pf-v5-c-empty-state__body"
    >
      <div>
        Image builder is a tool to compose customized RHEL (rpm-ostree) images optimized for Edge. 
        <br />
        <br />
      </div>
      <div>
        With OSTree, you can manage the system software by referencing a central image repository. Images contain a complete operating system ready to be remotely installed at scale. Updates to images are tracked through commits and enable secure updates that only address changes and keep the operating system unchanged. Image updates are quick, and rollbacks are easy.
        <br />
        <br />
      </div>
    </div>
    <div
      class="pf-v5-c-empty-state__footer"
    >
      <div
        class="pf-v5-c-empty-state__actions"
      >
        <div
          class="pf-v5-l-stack"
        >
          <div
            class="pf-v5-l-stack__item"
          >
            <a
              aria-disabled="false"
              class="pf-v5-c-button pf-m-link"
              data-ouia-component-id="OUIA-Generated-Button-link-1"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              href="https://ostreedev.github.io/ostree/"
              target="_blank"
            >
              Learn more about OSTree.
              <svg
                aria-hidden="true"
                class="pf-v5-svg pf-u-ml-sm"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                />
              </svg>
            </a>
          </div>
          <div
            class="pf-v5-l-stack__item"
          >
            <button
              aria-disabled="false"
              class="pf-v5-c-button pf-m-primary edge-stretched-button"
              data-ouia-component-id="OUIA-Generated-Button-primary-1"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              target=""
              type="button"
            >
              Create image.
            </button>
          </div>
          <div
            class="pf-v5-l-stack__item"
          >
            <a
              aria-disabled="false"
              class="pf-v5-c-button pf-m-link"
              data-ouia-component-id="OUIA-Generated-Button-link-2"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              href="https://access.redhat.com/documentation/en-us/edge_management/2022/html/create_rhel_for_edge_images_and_configure_automated_management/index"
              target="_blank"
            >
              Image builder for OSTree documentation
              <svg
                aria-hidden="true"
                class="pf-v5-svg pf-u-ml-sm"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                />
              </svg>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
